### PR TITLE
qa-lora integration

### PR DESCRIPTION
**Pull request by ved1beta**
# Description

qa-lora integration 

## Motivation and Context
#2878 

## How has this been tested?
run the ``` qalora.yml ``` file present 

## Types of changes

added 2 qalora specific parameters ``` qalora: bool ``` and ``` qalora_group_size: Optional[int] ```
also added qalora support in adapters loading and all 

inspiration [peft qalora](https://github.com/huggingface/peft/pull/2571) 

## Summary by CodeRabbit

- New Features
  - Added QALoRA adapter support across model loading, quantization, and training flows.
  - Provided an example QALoRA fine-tuning configuration.
  - Added a flash-attention QKV fusion option.

- Validation
  - Enforced QALoRA requirements (4-bit loading, required group size) and blocked incompatible merges.
  - Extended warnings and incompatibility checks for gradient checkpointing, fused attention, and distributed setups.
  - Added QALoRA enum option in configs.

- Tests
  - Expanded tests to cover QALoRA scenarios.

### CODE REVIEW

**Comment by winglian** [comment]
Were you able to merge the adapters back into the quantized model? If I understand the paper correctly, the benefit comes from the ability to merge it back into the quantized model.

**Comment by ved1beta** [comment]
as of now the PR Does not implement adapter merging back into quantized models ,

**Comment by ved1beta** [comment]
The PEFT library's QA-LoRA implementation intentionally doesn't support merging for GPTQ , i think we should keep it the same  #2878 we could just enhance the error messages

**Comment by NanoCode012** [comment]
Re: https://github.com/axolotl-ai-cloud/axolotl/issues/2878#issuecomment-3051834944

> Offering help with QA-LoRA adapter merge process! Since PEFT doesn't support adapter merging with quantized models yet, I've implemented a custom solution. Successfully replicated the QA-LoRA paper and built a working merge process for GPTQ + QA-LoRA adapters. Happy to share code/knowledge

@gapsong, could you share how you've implemented the merge?

**Comment by gapsong** [comment]
Hi @NanoCode012, @ved1beta, @winglian,

Thanks for the work on this. Here is the important change.
The code is a bit messy...but I hope you get the main idea.
https://github.com/ModelCloud/GPTQModel/compare/main...gapsong:GPTQModel:qzero_unquantized
I got the information from a conversation with the qa-lora author https://github.com/yuhuixu1993/qa-lora/issues/45

**Problem:**
I noticed the `qzero` values are currently being quantized during the save process.

**Why this is an issue:**
The subsequent merge of these `qzero` values is a high-precision operation. If we save them in a quantized format, we lose precision due to rounding *before* the merge even happens, which defeats the purpose.

**Solution:**
My changes address this by ensuring the `qzero` values are saved in a full-precision format (specifically `FP16`). This preserves their original values for the merge operation.

Happy to discuss this further!

File: src/axolotl/utils/schemas/peft.py [hunk 1]
```diff
@@ -128,6 +128,16 @@ class LoraConfig(BaseModel):
             "description": "loraplus learning rate for lora embedding layers. Default value is 1e-6."
         },
     )
+    use_qalora: bool | None = Field(
+        default=False,
+        json_schema_extra={
+            "description": "Enable Quantization-Aware Low-Rank Adaptation"
+        },
+    )
```
Comment (NanoCode012) [inline src/axolotl/utils/schemas/peft.py]:
```suggestion
```
This feels unneeded as we can specify via `adapter: use_qalora`. I think we can remove other places this is referenced too.

File: src/axolotl/utils/schemas/peft.py [hunk 2]
```diff
@@ -128,6 +128,16 @@ class LoraConfig(BaseModel):
...
+    qalora_group_size: int | None = Field(
+        default=16,
```
Comment (NanoCode012) [inline src/axolotl/utils/schemas/peft.py]: We may want default to be None, otherwise config will auto set this.
```suggestion
        default=None,
```

File: src/axolotl/utils/schemas/peft.py [hunk 3]
```diff
@@ -41,7 +41,7 @@ class LoraConfig(BaseModel):
     adapter: str | None = Field(
         default=None,
         json_schema_extra={
-            "description": "If you want to use 'lora' or 'qlora' or leave blank to train all parameters in original model"
+            "description": "If you want to use 'lora' or 'qlora' or 'qalora' or leave blank to train all parameters in original model"
         },
     )
```
Comment (NanoCode012) [inline src/axolotl/utils/schemas/peft.py]: Although unrelated, but I wonder if we can set `adapter: AdapterEnum | None`. This should simplify our checks later below to not allow incorrect string.

File: src/axolotl/utils/schemas/validation.py [hunk 4]
```diff
@@ -591,17 +591,28 @@ def check_lora_axolotl_unsloth(cls, data):
 
     @model_validator(mode="after")
     def check_fused_lora(self):
-        if self.adapter in ["lora", "qlora"] and (
+        if self.adapter in ["lora", "qlora", "qalora"] and (
             self.flash_attn_fuse_qkv or self.flash_attn_fuse_mlp
         ):
             raise ValueError("Fused modules are not supported with LoRA/QLoRA")
```
Comment (NanoCode012) [inline src/axolotl/utils/schemas/validation.py]:
```suggestion
            raise ValueError("Fused modules are not supported with LoRA/QLoRA/QALoRA")
```

File: src/axolotl/utils/schemas/validation.py [hunk 5]
```diff
@@ -591,17 +591,28 @@ def check_lora_axolotl_unsloth(cls, data):
...
+    @model_validator(mode="after")
+    def validate_qalora(self):
+        if self.adapter == "qalora":
+            if not self.load_in_4bit:
+                raise ValueError("QALoRA requires load_in_4bit to be True")
+            if not hasattr(self, "qalora_group_size") or self.qalora_group_size is None:
+                raise ValueError("QALoRA requires qalora_group_size to be specified")
+            if self.merge_lora:
+                raise ValueError("QALoRA does not support merge_lora yet")
+        return self
```
Comment (NanoCode012) [inline src/axolotl/utils/schemas/validation.py:592]: This feels like duplicate of validation in `src/axolotl/utils/schemas/peft.py`.

File: src/axolotl/loaders/adapter.py [hunk 6]
```diff
@@ -93,6 +93,13 @@ def load_lora(
     if loftq_bits:
         lora_config_kwargs["loftq_config"] = LoftQConfig(loftq_bits=loftq_bits)
         lora_config_kwargs["init_lora_weights"] = "loftq"
+
+    if cfg.adapter == "qalora":
+        if hasattr(cfg, "use_qalora") and cfg.use_qalora:
+            lora_config_kwargs["use_qalora"] = True
```
Comment (NanoCode012) [inline src/axolotl/loaders/adapter.py]: if the adapter is qalora, we can just set this lora kwarg to true.

File: src/axolotl/loaders/adapter.py [hunk 7]
```diff
@@ -93,6 +93,13 @@ def load_lora(
...
+        if hasattr(cfg, "qalora_group_size") and cfg.qalora_group_size:
+            lora_config_kwargs["qalora_group_size"] = cfg.qalora_group_size
```
Comment (NanoCode012) [inline src/axolotl/loaders/adapter.py:97]: Do you want to raise an Error here or possibly in validation if a user does not provide this?

File: src/axolotl/core/training_args_base.py [hunk 8]
```diff
@@ -163,6 +163,14 @@ class AxolotlTrainingMixins:
...
         metadata={"help": "whether this is a qlora training"},
     )
+    qalora: bool = field(
+        default=False,
+        metadata={"help": "whether this is a qalora training"},
+    )
+    qalora_group_size: Optional[int] = field(
+        default=16,
+        metadata={"help": "Group size for QALoRA quantization"},
+    )
```
Comment (NanoCode012) [inline src/axolotl/core/training_args_base.py]: I don't think this is needed in the training args.
```suggestion
```

File: src/axolotl/core/builders/causal.py [hunk 9]
```diff
@@ -151,6 +151,9 @@ def build(self, total_num_steps):
         if self.cfg.adapter == "qlora":
             training_arguments_kwargs["qlora"] = True
 
+        if self.cfg.adapter == "qalora":
+            training_arguments_kwargs["qalora"] = True
```
Comment (NanoCode012) [inline src/axolotl/core/builders/causal.py]: Might not be needed
```suggestion
```

Also similarly the above
```
        if self.cfg.adapter == "qlora":
            training_arguments_kwargs["qlora"] = True
```

seems unneeded too.

File: examples/llama-2/qalora.yml [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+
+# This will be the new adapter type once implemented
+adapter: qalora
+lora_model_dir:
+
+sequence_len: 2048
+sample_packing: true
+eval_sample_packing: false
+
+qlora: true
```
Comment (NanoCode012) [inline examples/llama-2/qalora.yml]:
```suggestion
```

**Review by NanoCode012 — COMMENTED** [review]
Thanks. I left some minor config points.

**Comment by NanoCode012** [comment]
@gapsong 

> I noticed the qzero values are currently being quantized during the save process.

Could you share where this is happening in peft?

**Comment by gapsong** [comment]
> @gapsong
> 
> > I noticed the qzero values are currently being quantized during the save process.
> 
> Could you share where this is happening in peft?
@NanoCode012 
It is not happening in peft it is happening in gptq. Either modelGPTQ or AutoGPTQ depending on the library that you are using.

File: src/axolotl/utils/schemas/validation.py [hunk 11]
```diff
@@ -574,6 +574,21 @@ def check_lora_axolotl_unsloth(cls, data):
 
     @model_validator(mode="after")
     def check_fused_lora(self):
+        if self.adapter in ["lora", "qlora", "qalora"] and (
+            self.flash_attn_fuse_qkv or self.flash_attn_fuse_mlp
+        ):
+            raise ValueError("Fused modules are not supported with LoRA/QLoRA/QALoRA")
+        return self
```
Comment (winglian) [inline src/axolotl/utils/schemas/validation.py]: I think this is a merge conflict as we removed `flash_attn_fuse_qkv` recently.
